File: 10-04/multi-language-ref/03-ext-libs/02-fftw3/fsynth.f90
! fourier synthesis code example in fortran 95

MODULE kinds
  IMPLICIT NONE
  SAVE
  INTEGER, PARAMETER :: DP = selected_real_kind(14,200)
END MODULE kinds

SUBROUTINE write_fsynth(freq,phase,intens,fnum,tnum,tmax,fname)
  USE kinds
  IMPLICIT NONE

  CHARACTER(len=*),INTENT(IN) :: fname
  INTEGER, INTENT(IN)         :: fnum,tnum
  REAL(kind=DP),INTENT(IN)    :: freq(1:*),phase(1:*),intens(1:*)
  REAL(kind=DP),INTENT(IN)    :: tmax
  REAL(kind=DP)               :: tdelta,t,val
  REAL(kind=DP),PARAMETER     :: twopi=6.28318530717958647688_DP
  INTEGER :: i,j

  tdelta=tmax/DBLE(tnum)

  OPEN(unit=9,file=TRIM(fname),status='unknown',form='formatted')

  WRITE(9,'(A,I10)') '# ', tnum

  DO i=0,tnum-1
     t=DBLE(i)*tdelta
     val=0.0d0

     DO j=1,fnum
        val=val+intens(j)*SIN(twopi*(freq(j)*t+phase(j)))
     END DO

     WRITE(9,*) t, val
  END DO

  CLOSE(9)

END SUBROUTINE write_fsynth


PROGRAM fsynth
  USE kinds

  IMPLICIT NONE
  INTEGER                   :: fnum,tnum,i
  REAL(kind=DP),ALLOCATABLE :: freq(:),phase(:),intens(:)
  REAL(kind=DP)             :: tmax
  CHARACTER(len=80)         :: fname

  PRINT*,'fourier synthesis'
  PRINT*,'give number of frequencies'
  READ*,fnum
  PRINT*,'got: ',fnum

  ALLOCATE(freq(fnum),phase(fnum),intens(fnum))

  DO i=1,fnum
     PRINT*,'frequency number: ',i
     PRINT*,'give frequency, intensity and phase:'
     READ*,freq(i),intens(i),phase(i)
     PRINT*,'got: ',freq(i),intens(i),phase(i)
  END DO

  PRINT*,'give total time'
  READ*,tmax
  PRINT*,'got: ',tmax
  PRINT*,'give number of intervals'
  READ*,tnum
  PRINT*,'got: ',tnum
  PRINT*,'give filename'
  READ(5,'(A)') fname
  PRINT*,'got: ',TRIM(fname)

  PRINT*,'generating output'
  CALL write_fsynth(freq,phase,intens,fnum,tnum,tmax,TRIM(fname))

!  DEALLOCATE(freq,phase,intens)
      
END PROGRAM fsynth

File: 10-04/multi-language-ref/03-ext-libs/02-fftw3/fsynth.c
/* fourier synthesis code example in C */

#include <stdio.h>
#include <math.h>

#include "utilities.h"

static const double TWOPI=6.28318530717958647688;

static void write_fsynth(char *fname, double *freq, double *phase,
                         double *intens, int fnum, int tnum, double tmax)
{
    int i,j;
    double tdelta,t,val;
    FILE *fp;

    tdelta = tmax / (double)tnum;

    fp = fopen(fname,"w");
    fprintf(fp,"# %d\n", tnum);

    for (i=0; i < tnum; ++i) {
        t = ((double) i) * tdelta;
        val = 0.0;

        for (j=0; j < fnum; ++j) {
            val += intens[j]*sin(TWOPI*(freq[j]*t+phase[j]));
        }
        fprintf(fp,"%lg %lg\n", t, val);
    }
    fclose(fp);
}

#define FMAX 20

int main(int argc, char **argv)
{
    int fnum, tnum, i;
    double freq[FMAX],phase[FMAX],intens[FMAX],tmax;
    char line[256],fname[81];

    printf("fourier synthesis\n");
    printf("give number of frequencies\n");
    fgets(line,255,stdin);
    sscanf(line,"%d",&fnum);
    printf("got: %d\n",fnum);

    if ((fnum < 1) || (fnum > FMAX)) {
        printf("only 1-%d fequencies are supported\n",FMAX);
        return 1;
    }

    for (i=0; i < fnum; ++i) {
        printf("frequency number: %d\n",i);
        printf("give frequency, intensity and phase:\n");
        fgets(line,255,stdin);
        sscanf(line,"%lg%lg%lg",&freq[i],&intens[i],&phase[i]);
        printf("got: %lg %lg %lg\n",freq[i],intens[i],phase[i]);
    }
       
    printf("give total time\n");
    fgets(line,255,stdin);
    sscanf(line,"%lg",&tmax);
    printf("got: %lg\n",tmax);

    printf("give number of intervals\n");
    fgets(line,255,stdin);
    sscanf(line,"%d",&tnum);
    printf("got: %d\n",tnum);

    printf("give filename\n");
    fgets(line,255,stdin);
    sscanf(line,"%s",fname);
    printf("got: %s\n",fname);

    printf("generating output\n");
    start_timer();
    write_fsynth(fname,freq,phase,intens,fnum,tnum,tmax);
    printf("time for generating output: %g seconds\n",
	   get_timer());

#if 0
    init_random(123455);
    for (i=0; i < 20; ++i)
      printf("%d: %12.8f\n", i, get_random());
#endif

    return 0;
}


